File: backend/app/schemas/startup.py
from typing import List, Optional, Union, Dict, Any
from pydantic import BaseModel, Field

class StartupAnalysis(BaseModel):
    summary: str = Field(..., description="Executive teardown of the concept and core value")
    problem: str = Field(..., description="Problem statement and friction points")
    solution: str = Field(..., description="Proposed solution and architecture")
    innovation: str = Field(..., description="Technical and market innovation tier")
    strengths: List[str] = Field(default_factory=list, description="Validated strong points")
    weaknesses: List[str] = Field(default_factory=list, description="Vulnerabilities and execution bottlenecks")
    risks: List[str] = Field(default_factory=list, description="Strategic and regulatory risks")
    suggestions: List[str] = Field(default_factory=list, description="Actionable analyst recommendations")
    opportunity_score: int = Field(..., ge=0, le=100, description="Numerical score 0-100")
    verdict: str = Field(..., description="Core assessment summary")

class CompetitorDetail(BaseModel):
    name: str
    focus: str
    advantage: str
    disadvantage: str

class MarketResearch(BaseModel):
    target_customers: List[str] = Field(default_factory=list)
    customer_segments: List[str] = Field(default_factory=list)
    market_size: str = Field(..., description="TAM / SAM / SOM addressable breakdown")
    market_demand: str = Field(..., description="Demand growth trajectory and CAGR")
    industry_trends: List[str] = Field(default_factory=list)
    competitors: Union[List[str], List[CompetitorDetail]] = Field(default_factory=list)
    customer_pain_points: List[str] = Field(default_factory=list)
    market_opportunities: List[str] = Field(default_factory=list)
    market_challenges: List[str] = Field(default_factory=list)

class BusinessStrategy(BaseModel):
    business_model: str
    revenue_model: str
    pricing_strategy: str
    unique_value_proposition: str
    go_to_market_strategy: str
    customer_acquisition: List[str] = Field(default_factory=list)
    sales_channels: List[str] = Field(default_factory=list)
    partnerships: List[str] = Field(default_factory=list)
    growth_strategy: str
    # Optional 9-block Canvas attributes
    key_partners: Optional[List[str]] = None
    key_activities: Optional[List[str]] = None
    key_resources: Optional[List[str]] = None
    customer_relationships: Optional[List[str]] = None
    cost_structure: Optional[List[str]] = None

class FinancialYearProjection(BaseModel):
    year: str
    revenue: float
    costs: float
    gross_profit: float

class FinancialPlan(BaseModel):
    startup_cost_estimate: str
    operating_costs: str
    revenue_projection: str
    break_even_estimate: str
    funding_required: str
    funding_utilization: str
    financial_risks: List[str] = Field(default_factory=list)
    profitability_potential: str
    timeline_projections: Optional[List[FinancialYearProjection]] = None

class SWOTAnalysis(BaseModel):
    strengths: List[str] = Field(default_factory=list)
    weaknesses: List[str] = Field(default_factory=list)
    opportunities: List[str] = Field(default_factory=list)
    threats: List[str] = Field(default_factory=list)

class InvestmentReport(BaseModel):
    executive_summary: str
    elevator_pitch: str
    swot: SWOTAnalysis
    investment_readiness_score: int = Field(..., ge=0, le=100)
    recommended_funding_stage: str = Field(..., description="Pre-seed | Seed | Series A | Not investment ready")
    investor_recommendation: str
    pitch_deck_outline: List[str] = Field(default_factory=list)
    final_verdict: str

class StartupReport(BaseModel):
    id: Optional[str] = None
    idea: str
    created_at: Optional[str] = None
    startup_analysis: StartupAnalysis
    market_research: MarketResearch
    business_strategy: BusinessStrategy
    financial_plan: FinancialPlan
    investment_report: InvestmentReport
    source: Optional[str] = "api"
    optional_inputs: Optional[Dict[str, Any]] = None

class AnalyzeStartupRequest(BaseModel):
    idea: str
    industry: Optional[str] = None
    target_customer: Optional[str] = None
    region: Optional[str] = None
    stage: Optional[str] = None
    budget: Optional[str] = None

File: src/types/startup.ts
/**
 * TypeScript Data Models for AI Startup Validator
 * Exactly mirrors the multi-agent backend output specification.
 */

export interface StartupAnalysis {
  summary: string;
  problem: string;
  solution: string;
  innovation: string;
  strengths: string[];
  weaknesses: string[];
  risks: string[];
  suggestions: string[];
  opportunity_score: number;
  verdict: string;
}

export interface CompetitorDetail {
  name: string;
  focus: string;
  advantage: string;
  disadvantage: string;
}

export interface MarketResearch {
  target_customers: string[];
  customer_segments: string[];
  market_size: string;
  market_demand: string;
  industry_trends: string[];
  competitors: string[] | CompetitorDetail[];
  customer_pain_points: string[];
  market_opportunities: string[];
  market_challenges: string[];
}

export interface BusinessStrategy {
  business_model: string;
  revenue_model: string;
  pricing_strategy: string;
  unique_value_proposition: string;
  go_to_market_strategy: string;
  customer_acquisition: string[];
  sales_channels: string[];
  partnerships: string[];
  growth_strategy: string;
  // Optional derived Business Model Canvas blocks
  key_partners?: string[];
  key_activities?: string[];
  key_resources?: string[];
  customer_relationships?: string[];
  cost_structure?: string[];
}

export interface FinancialYearProjection {
  year: string;
  revenue: number;
  costs: number;
  gross_profit: number;
}

export interface FinancialPlan {
  startup_cost_estimate: string;
  operating_costs: string;
  revenue_projection: string;
  break_even_estimate: string;
  funding_required: string;
  funding_utilization: string;
  financial_risks: string[];
  profitability_potential: string;
  // Optional numerical timeline projection for plotting
  timeline_projections?: FinancialYearProjection[];
}

export interface SWOTAnalysis {
  strengths: string[];
  weaknesses: string[];
  opportunities: string[];
  threats: string[];
}

export interface InvestmentReport {
  executive_summary: string;
  elevator_pitch: string;
  swot: SWOTAnalysis;
  investment_readiness_score: number;
  recommended_funding_stage: 'Pre-seed' | 'Seed' | 'Series A' | 'Not investment ready' | string;
  investor_recommendation: string;
  pitch_deck_outline: string[];
  final_verdict: string;
}

export interface StartupReport {
  id?: string;
  idea: string;
  created_at?: string;
  startup_analysis: StartupAnalysis;
  market_research: MarketResearch;
  business_strategy: BusinessStrategy;
  financial_plan: FinancialPlan;
  investment_report: InvestmentReport;
  source?: 'api' | 'demo';
  optional_inputs?: {
    industry?: string;
    target_customer?: string;
    region?: string;
    stage?: string;
    budget?: string;
  };
}

export interface AnalyzeStartupRequest {
  idea: string;
  industry?: string;
  target_customer?: string;
  region?: string;
  stage?: string;
  budget?: string;
}

export type AgentStageId = 1 | 2 | 3 | 4 | 5;

export type AgentStatus = 'waiting' | 'running' | 'completed' | 'failed';

export interface AgentStageInfo {
  id: AgentStageId;
  number: string;
  name: string;
  role: string;
  purpose: string;
  outputDescription: string;
  status: AgentStatus;
  startedAt?: number;
  completedAt?: number;
  errorMessage?: string;
  logMessage?: string;
}

export interface AnalysisHistoryItem {
  id: string;
  idea: string;
  opportunity_score: number;
  investment_readiness_score: number;
  verdict: string;
  funding_stage: string;
  date: string;
  createdAt?: string;
  status?: string;
  report: StartupReport;
}

export type HistoryItem = AnalysisHistoryItem;

